Add event aggregation use case to Data Prepper documentation (#6206)
opensearch-project/documentation-website: _data-prepper/common-use-cases/event-aggregation.md

---
layout: default

> title: Event aggregation with Data Prepper

Reviewer: Can we also drop the " with Data Prepper" here?

parent: Common use cases
nav_order: 40
---

# Event aggregation with Data Prepper

> You can use Data Prepper to aggregate data from different events over a period of time. Aggregating events can help reduce unnecessary log volume and handle use cases like multiline logs that come in as separate events. The [Aggregate processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/aggregate/) is a stateful processor that groups events based on the values for a set of specified identification keys, and performs a configurable action on each group.

Reviewer: I think it may be advantageous to format the aggregate processor as [`aggregate` processor](https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/processors/aggregate/) or [`aggregate`](https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/processors/aggregate/) processor.

Thus, change:

```
The [Aggregate processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/aggregate/)
```

to

```
The [`aggregate` processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/aggregate/)
```

State in the Aggregate processor is stored in memory. For example, in order to combine four events into one, the processor needs to retain pieces of the first three events. The state of an aggregate group of events is kept for a configurable amount of time. Depending on your logs, the aggregate action being used, and the amount of memory options in the processor configuration, the aggregation could take place over a long period of time.

## Basic usage

> The following example pipeline extracts the fields `sourceIp`, `destinationIp`, and `port` using the [Grok processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/grok/), and then aggregates on those fields over a period of 30 seconds using the [Aggregate processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/aggregate/) and the `put_all` action. At the end of the 30 seconds, the aggregated log is sent to the OpenSearch sink.

Reviewer: Similar comment as above with [Aggregate processor].

Author: I'll carryover this change across the other use case PRs.

```json
aggregate_pipeline:  
   source:
     http:
      path: "/${pipelineName}/logs"
   processor:
     - grok:
         match: 
           log: ["%{IPORHOST:sourceIp} %{IPORHOST:destinationIp} %{NUMBER:port:int}"]
     - aggregate:
         group_duration: "30s"
         identification_keys: ["sourceIp", "destinationIp", "port"]
         action:
           put_all:
   sink:
     - opensearch:
         ...
         index: aggregated_logs
```
{% include copy-curl.html %}

For example, consider the following batch of logs:

```json
{ "log": "127.0.0.1 192.168.0.1 80", "status": 200 }
{ "log": "127.0.0.1 192.168.0.1 80", "bytes": 1000 }
{ "log": "127.0.0.1 192.168.0.1 80" "http_verb": "GET" }
```
{% include copy-curl.html %}

> The Grok processor will extract the `identification_keys` to create the following logs:

Reviewer: The grok processor is not creating the following logs. It is modifying the existing log events to look like what follows. Also, it is extract keys more generally. The pipeline author has to extract them to match what `aggregate` expects for `identification_keys`.

```
The grok processor will extract extract keys such that the log events look like the following. These events now have the data which the aggregate processor will need for the `identification_keys`.
```

```json
{ "sourceIp": "127.0.0.1", "destinationIp": "192.168.0.1", "port": 80, "status": 200, "bytes": 1000, "http_verb": "GET" }
```
{% include copy-curl.html %}

## Removing duplicates

You can remove duplicate entries by deriving keys from incoming events and specifying the `remove_duplicates` option for the `aggregate` processor. This action immediately processes the first event for a group, and drops all following events in that group.

In the following example, the first event is processed with the identification keys `sourceIp` and `destinationIp`:

```json
{ "sourceIp": "127.0.0.1", "destinationIp": "192.168.0.1", "status": 200 }
```
{% include copy-curl.html %}

The pipeline will then drop the following event because it has the same keys:

```json
{ "sourceIp": "127.0.0.1", "destinationIp": "192.168.0.1", "bytes": 1000 }
```
{% include copy-curl.html %}

The pipeline processes this event and creates a new group because the `sourceIp` is different:

```json
{ "sourceIp": "127.0.0.2", "destinationIp": "192.168.0.1", "bytes": 1000 }
```
{% include copy-curl.html %}

## Log aggregation and conditional routing

> You can use multiple plugins to combine log aggregation with conditional routing. In this example, the sub-pipeline `log-aggregate-pipeline` receives logs by using an HTTP client like FluentBit and extracts important values from the logs by matching the value in the `log` key against the common Apache log pattern.

Reviewer 2: ```suggestion
You can use multiple plugins to combine log aggregation with conditional routing. In this example, the sub-pipeline `log-aggregate-pipeline` receives logs by using an HTTP client, like FluentBit, and extracts important values from the logs by matching the value in the `log` key against the common Apache log pattern.
```

> You can use multiple plugins to combine log aggregation with conditional routing. In this example, the sub-pipeline `log-aggregate-pipeline` receives logs by using an HTTP client like FluentBit and extracts important values from the logs by matching the value in the `log` key against the common Apache log pattern.

Reviewer 2: End of last sentence: "against the Apache Common Log Format"?

> Two of the values it extracts from the logs with a grok pattern include `response` and `clientip`. The Aggregate processor then uses the `clientip` value, along with the `remove_duplicates` option, to drop any logs that contain a `clientip` that has already been processed within the given `group_duration`.

Reviewer 3: ```suggestion
Two of the values the sub-pipeline extracts from the logs with a grok pattern include `response` and `clientip`. The Aggregate processor then uses the `clientip` value, along with the `remove_duplicates` option, to drop any logs that contain a `clientip` that has already been processed within the given `group_duration`.
```

> Two of the values the sub-pipeline extracts from the logs with a grok pattern include `response` and `clientip`. The `aggregate` processor then uses the `clientip` value, along with the `remove_duplicates` option, to drop any logs that contain a `clientip` that has already been processed within the given `group_duration`.

Reviewer 2: ```suggestion
Two of the values that the sub-pipeline extracts from the logs with a grok pattern include `response` and `clientip`. The `aggregate` processor then uses the `clientip` value, along with the `remove_duplicates` option, to drop any logs that contain a `clientip` that has already been processed within the given `group_duration`.
```

> Two of the values the sub-pipeline extracts from the logs with a grok pattern include `response` and `clientip`. The `aggregate` processor then uses the `clientip` value, along with the `remove_duplicates` option, to drop any logs that contain a `clientip` that has already been processed within the given `group_duration`.

Reviewer 2: Should "Grok" be capitalized?

Author: It looks like we capitalize when it's part of a proper name, such as Grok Debugger, lowercase it in common usage.

Reviewer 2: It looks like "Grok pattern" is used [here](https://streamsets.com/blog/what-are-grok-patterns/) and [here](https://graylog.org/post/getting-started-with-grok-patterns/) and in many other places, with Elastic being the notable exception. I would lean toward capitalizing it if it's a reference to a proper noun, which it appears to be.

Author: Okay. I'll clean up the capitalization across Data Prepper and OpenSearch ingest processors.

> Three routes, or conditional statements, exist in the pipeline. These routes separate the value of the response into 2xx/3xx, 4xx, and 5xx responses. Logs with a 2xx and 3xx status are sent to the aggregated_2xx_3xx index, logs with a 4xx status are sent to the aggregated_4xx index, and logs with a 5xx status are sent to the aggregated_5xx index.

Reviewer 2: Second sentence: Confirm that it shouldn't be "2xx or 3xx status". Should the index names be in code font?

Reviewer: @natebower  is correct. These are logs that have either a 2xx or 3xx status.

Reviewer: @vagimeli , We want to change this:

> Logs with a `2xx` and `3xx` status

to:

> Logs with a `2xx` or `3xx` status
